CFlow inference error

It may exists an error in cflow inference:
in models/cflow/anomaly_map.py ,line 59 is :
            test_map.append(
                F.interpolate(
                    test_mask.unsqueeze(1), size=self.image_size, mode="bilinear", align_corners=True
                ).squeeze()
at train steps, the test_mask shape is [batch_size,32,32] (batch_size>1):
1、after F.interpolate() ,it becomes [batch_size,1,image_size,image_size] 
2、after squeeze(),it becomes [batch_size,image_size,image_size]
when inference,for one image ,the test_mask shape is [1,32,32]：
1、after F.interpolate() ,it becomes [1,1,image_size,image_size] 
2、but after squeeze(),it becomes [image_size,image_size]
so, I think the squeeze() should be squeeze(1) , make length of shape be 3 which is [1,image_size,image_size].

@samet-akcay Hi,thanks your reply,this is my config file:
"
dataset:
  name: mvtec #options: [mvtec, btech, folder]
  format: mvtec
  path: /media/ycao/data0/anomalib/mvtec
  category: toothbrush
  task: segmentation
  image_size: 256
  train_batch_size: 16
  test_batch_size: 16
  inference_batch_size: 16
  fiber_batch_size: 64
  num_workers: 0
  transform_config:
    train: null
    val: null
  create_validation_set: false

model:
  name: cflow
  backbone: resnet18
  pre_trained: true
  layers:
    - layer2
    - layer3
    - layer4
  decoder: freia-cflow
  condition_vector: 128
  coupling_blocks: 8
  clamp_alpha: 1.9
  soft_permutation: false
  lr: 0.0001
  early_stopping:
    patience: 2
    metric: pixel_AUROC
    mode: max
  normalization_method: min_max # options: [null, min_max, cdf]

metrics:
  image:
    - F1Score
    - AUROC
  pixel:
    - F1Score
    - AUROC
  threshold:
    image_default: 0
    pixel_default: 0
    adaptive: true

visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: True # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]
...
"
but I think the error has nothing to do with config file.
when I run lightning_inference.py, the  batch_size in  "DataLoader(dataset)"  defaults to 1, then i got an error:
  File "/home/ycao/.conda/envs/d2/lib/python3.9/site-packages/anomalib/post_processing/post_process.py", line 131, in superimpose_anomaly_map
    superimposed_map = cv2.addWeighted(anomaly_map, alpha, image, (1 - alpha), gamma)
cv2.error: OpenCV(4.5.5) /io/opencv/modules/core/src/arithm.cpp:647: error: (-209:Sizes of input arguments do not match) The operation is neither 'array op array' (where arrays have the same size and the same number of channels), nor 'array op scalar', nor 'scalar op array' in function 'arithm_op'

so I traced back to  anomaly_map.py, after I modified " squeeze()" to " squeeze(1)", the error has gone !

--- a/anomalib/models/cflow/anomaly_map.py
+++ b/anomalib/models/cflow/anomaly_map.py
@@ -24,9 +24,7 @@ class AnomalyMapGenerator(nn.Module):
         self.image_size = image_size if isinstance(image_size, tuple) else tuple(image_size)
         self.pool_layers: List[str] = pool_layers
 
-    def compute_anomaly_map(
-        self, distribution: Union[List[Tensor], List[List]], height: List[int], width: List[int]
-    ) -> Tensor:
+    def compute_anomaly_map(self, distribution: List[Tensor], height: List[int], width: List[int]) -> Tensor:
         """Compute the layer map based on likelihood estimation.
 
         Args:
@@ -38,23 +36,22 @@ class AnomalyMapGenerator(nn.Module):
           Final Anomaly Map
 
         """
-
-        test_map: List[Tensor] = []
+        layer_maps: List[Tensor] = []
         for layer_idx in range(len(self.pool_layers)):
-            test_norm = torch.tensor(distribution[layer_idx], dtype=torch.double)  # pylint: disable=not-callable
-            test_norm -= torch.max(test_norm)  # normalize likelihoods to (-Inf:0] by subtracting a constant
-            test_prob = torch.exp(test_norm)  # convert to probs in range [0:1]
-            test_mask = test_prob.reshape(-1, height[layer_idx], width[layer_idx])
+            layer_distribution = distribution[layer_idx].clone().detach()
+            # Normalize the likelihoods to (-Inf:0] and convert to probs in range [0:1]
+            layer_probabilities = torch.exp(layer_distribution - layer_distribution.max())
+            layer_map = layer_probabilities.reshape(-1, height[layer_idx], width[layer_idx])
             # upsample
-            test_map.append(
+            layer_maps.append(
                 F.interpolate(
-                    test_mask.unsqueeze(1), size=self.image_size, mode="bilinear", align_corners=True
-                ).squeeze()
+                    layer_map.unsqueeze(1), size=self.image_size, mode="bilinear", align_corners=True
+                ).squeeze(1)
             )
         # score aggregation
-        score_map = torch.zeros_like(test_map[0])
+        score_map = torch.zeros_like(layer_maps[0])
         for layer_idx in range(len(self.pool_layers)):
-            score_map += test_map[layer_idx]
+            score_map += layer_maps[layer_idx]
         score_mask = score_map
         # invert probs to anomaly scores
         anomaly_map = score_mask.max() - score_mask

--- a/anomalib/models/cflow/anomaly_map.py
+++ b/anomalib/models/cflow/anomaly_map.py
@@ -52,9 +52,9 @@ class AnomalyMapGenerator(nn.Module):
         score_map = torch.zeros_like(layer_maps[0])
         for layer_idx in range(len(self.pool_layers)):
             score_map += layer_maps[layer_idx]
-        score_mask = score_map
-        # invert probs to anomaly scores
-        anomaly_map = score_mask.max() - score_mask
+
+        # Invert probs to anomaly scores
+        anomaly_map = score_map.max() - score_map
 
         return anomaly_map
 

